### PR TITLE
 Panic in filter

### LINKED ISSUE #459: Panic in filter
```
goroutine 91 [running:
github.com/vova616/xxhash.Checksum32Seed(0xc002966050, 0x0, 0x0, 0xc000000000, 0xc0bd7378e6)
    /go/pkg/mod/github.com/vova616/xxhash@v0.0.0-20130313230233-f0a9a8b74d48/xxhash.go:189 +0x19d
github.com/vova616/xxhash.Checksum32(...)
    /go/pkg/mod/github.com/vova616/xxhash@v0.0.0-20130313230233-f0a9a8b74d48/xxhash.go:181
github.com/moira-alert/moira/filter.findPart(0xc002966050, 0x0, 0xc002314e00, 0x1, 0x40, 0xc002314e00, 0x1, 0x40, 0x1)
    /go/src/github.com/moira-alert/moira/filter/pattern_index.go:124 +0xd8
github.com/moira-alert/moira/filter.(*PatternIndex).MatchPatterns(0xc0026f1ce0, 0xc002966050, 0x23, 0xf4e4d0, 0xc002966050, 0xc0025789c0)
    /go/src/github.com/moira-alert/moira/filter/pattern_index.go:107 +0x2f3
github.com/moira-alert/moira/filter.(*PatternStorage).matchPatterns(0xc00010b7c0, 0xc0026f9740, 0xf315c0, 0xc0026f9740, 0x0)
    /go/src/github.com/moira-alert/moira/filter/patterns_storage.go:91 +0x11c
github.com/moira-alert/moira/filter.(*PatternStorage).ProcessIncomingMetric(0xc00010b7c0, 0xc002966050, 0x40, 0x41, 0x0)
    /go/src/github.com/moira-alert/moira/filter/patterns_storage.go:68 +0x1eb
github.com/moira-alert/moira/filter/patterns.(*Matcher).worker(0xc000108cb0, 0xc001a00300, 0xc001a00360, 0x0, 0x0)
    /go/src/github.com/moira-alert/moira/filter/patterns/matcher.go:51 +0x92
github.com/moira-alert/moira/filter/patterns.(*Matcher).Start.func1(0x0, 0x0)
    /go/src/github.com/moira-alert/moira/filter/patterns/matcher.go:35 +0x3c
gopkg.in/tomb%!e(MISSING)v2.(*Tomb).run(0xc000108cc0, 0xc001485660)
    /go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:163 +0x2b
created by gopkg.in/tomb%!e(MISSING)v2.(*Tomb).Go
    /go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:159 +0xc7
panic: runtime error: index out of range [0 with length 0
```

### PR DESCRIPTION
**Pull request by Nixolay**
Fix #459

### CODE REVIEW

File: filter/pattern_index.go [hunk 1]
```diff
@@ -7,16 +7,17 @@ import (
 
 	"github.com/moira-alert/moira"
 
-	"github.com/vova616/xxhash"
+	//"github.com/vova616/xxhash"
```
Comment (litleleprikon) [inline filter/pattern_index.go]: Please remove commented code

File: filter/pattern_index.go [hunk 2]
```diff
@@ -57,7 +58,8 @@ func NewPatternIndex(logger moira.Logger, patterns []string) *PatternIndex {
 				}
 
 				if part == "*" || !strings.ContainsAny(part, "{*?") {
-					newNode.Hash = xxhash.Checksum32([]byte(part))
+					//newNode.Hash = xxhash.Checksum32([]byte(part))
```
Comment (litleleprikon) [inline filter/pattern_index.go]: Please remove commented code

**Comment by Nixolay** [comment]
image
The code only checks for emptiness, while simulating with the test data of an empty array, it gave a similar picture that we had.
```
func Checksum32(data []byte) uint32 {
	return Checksum32Seed(data, 0)
}

// Checksum32 returns the xxhash32 checksum of data. Length of data MUST BE less than 2 Gigabytes.
func Checksum32Seed(data []byte, seed uint32) uint32 {
	if data == nil {
		panic("Data cannot be nil.")
	}
```

File: go.mod [hunk 3]
```diff
@@ -20,6 +19,7 @@ require (
 	github.com/boltdb/bolt v1.3.2-0.20180302180052-fd01fc79c553 // indirect
 	github.com/bwmarrin/discordgo v0.19.1-0.20190513173526-8d8906ce4b95
 	github.com/carlosdp/twiliogo v0.0.0-20161027183705-b26045ebb9d1
+	github.com/cespare/xxhash/v2 v2.1.1
```
Comment (beevee) [inline go.mod]: I expected old library to disappear from dependencies, but it did not. Maybe run `go mod tidy` or something.

**Comment by litleleprikon** [comment]
Can we cover the nil pointer dereference with tests that will show us that bug is fixed?

**Comment by beevee** [comment]
@litleleprikon we have replaced `xxhash.Checksum32(moira.UnsafeStringToBytes(part))` with `xxhash.Sum64String(part)`, bypassing our potentially flawed code entirely. I don't see the point in testing third-party libraries.

**Comment by litleleprikon** [comment]
@beevee I meant that library is failing on our data and probably test that new library does not failing on our data will be helpful. But writing this test will require extra work from us and of course if you don't thing that this is necessary we can skip it.

File: unsafe.go [hunk 4]
```diff
@@ -16,10 +16,11 @@ func UnsafeBytesToString(b []byte) string {
 
 // UnsafeStringToBytes converts string to source without copying
 func UnsafeStringToBytes(s string) []byte {
```
Comment (beevee) [inline unsafe.go]: Wait, why do we need this code, exactly? @Nixolay you've said that it is still used somewhere, but I can't find any usages except in tests that test this code.

Comment (beevee) [reply unsafe.go]: OK, my bad, ignore this comment.

**Pull request merged**